src/modules/qt/filter_typewriter.cpp

```diff
@@ -0,0 +1,313 @@
+        producer = &kt->parent;
+        producer_properties = MLT_PRODUCER_PROPERTIES( producer );
+
+        if (producer == nullptr || producer_properties == nullptr)
+            return 0;
+
+        d = mlt_properties_get( producer_properties, "resource" );
+        cont->is_template = (d && d[0] != '\0');
+
+        if (cont->is_template)
+            d = mlt_properties_get( producer_properties, "_xmldata");
+        else
+            d = mlt_properties_get( producer_properties, "xmldata" );
+
+        if (d == nullptr)
+            return 0;
+
+        step_length = mlt_properties_get_int(filter_p, "step_length");
+        sigma = mlt_properties_get_int(filter_p, "step_sigma");
+        seed = mlt_properties_get_int(filter_p, "random_seed");
+        macro = mlt_properties_get_int(filter_p, "macro_type");
+
+        // if xml data changed, set update mask 0x1
+        if (cont->xml_data != d || macro != cont->macro)
+            update_mask = 0x3;
+
+        if (step_length != cont->step_length || sigma != cont->sigma || seed != cont->seed)
+            update_mask |= 0x2;
+
+        // clear and prepare for new parsing
+        if (0 == update_mask)
+            return 1;
+    }
+    else
+    {
+        return 0;
+    }
+
+    if (update_mask & 0x1)
+    {
+        cont->clean();
+
+        // save new data field name
+        cont->xml_data = d;
+
+        // Get content data and backup in the tw container.
+        cont->xp.setDocument(d);
+        cont->xp.parse();
+        unsigned int n = cont->xp.getContentNodesNumber();
+        for (uint i = 0; i < n; ++i)
+        {
+            std::string key = cont->xp.getNodeContent(i).toStdString();
+            TypeWriter data;
+
+            if (macro) {
+                char * buff = new char[key.length()+5];
+                char c = 0;
+                switch (macro) {
+                    case 1: c = 'c'; break;
+                    case 2: c = 'w'; break;
+                    case 3: c = 'l'; break;
+                    default: break;
+                }
+
+                sprintf(buff, ":%c{%s}", c, key.c_str());
+                data.setPattern(buff);
+                delete [] buff;
+            } else {
+                data.setPattern(key);
+            }
+            cont->renders.push_back(data);
+        }
+
+        cont->macro = macro;
+        cont->producer_type = 1;
+        cont->producer = producer;
+
+        // mark as inited
+        cont->init = true;
+    }
+
+    if (update_mask & 0x2)
+    {
+        for (auto & render : cont->renders)
+        {
+            render.setFrameStep(step_length);
+            render.setStepSigma(sigma);
+            render.setStepSeed(seed);
+            render.parse();
+        }
+        cont->step_length = step_length;
+        cont->sigma = sigma;
+        cont->seed = seed;
+    }
+
+    return 1;
+}
+
+static int update_producer(mlt_frame frame, mlt_properties frame_p, FilterContainer * cont, int restore)
```

Comment:
Parameter `frame_p` is not used and can be removed. Also, restore can be a `bool`.

Reply:
This and others unused arguments has been hide (commented out).